samples/core/Querying/NullSemantics/Program.cs

```diff
@@ -0,0 +1,80 @@
+﻿using Microsoft.EntityFrameworkCore;
+using System;
+using System.Linq;
+using System.Linq.Expressions;
+
+namespace NullSemantics
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            using var context = new NullSemanticsContext();
+            context.Database.EnsureDeleted();
+            context.Database.EnsureCreated();
+
+            #region FunctionSqlRaw
+            context.Database.ExecuteSqlRaw(
+                @"create function [dbo].[ConcatStrings] (@prm1 nvarchar(max), @prm2 nvarchar(max))
+                    returns nvarchar(max)
+                    as
+                    begin
+                        return @prm1 + @prm2;
+                    end");
+            #endregion
+
+            BasicExamples();
+            Functions();
+            ManulaOptimization();
+        }
+
+        static void BasicExamples()
+        {
+            using var context = new NullSemanticsContext();
+            #region BasicExamples
+            var query1 = context.Entities.Where(e => e.Id == e.Int);
+            var query2 = context.Entities.Where(e => e.Id == e.NullableInt);
+            var query3 = context.Entities.Where(e => e.Id != e.NullableInt);
+            var query4 = context.Entities.Where(e => e.String1 == e.String2);
+            var query5 = context.Entities.Where(e => e.String1 != e.String2);
+            #endregion
+
+            var result1 = query1.ToList();
+            var result2 = query2.ToList();
+            var result3 = query3.ToList();
+            var result4 = query4.ToList();
+            var result5 = query5.ToList();
+        }
+
+        static void Functions()
+        {
+            using var context = new NullSemanticsContext();
+
+            #region Functions
+            var query = context.Entities.Where(e => e.String1.Substring(0, e.String2.Length) == null);
+            #endregion
+
+            #region UdfExamples
+            var query1 = context.Entities.Where(e => context.ConcatStrings(e.String1, e.String2) != null);
+            var query2 = context.Entities.Where(e => context.ConcatStringsOptimized(e.String1, e.String2) != null);
+            #endregion
+
+            var result = query.ToList();
+            var result1 = query1.ToList();
+            var result2 = query2.ToList();
+        }
+
+        static void ManulaOptimization()
+        {
+            using var context = new NullSemanticsContext();
+
+            #region ManualOptimization
+            var query1 = context.Entities.Where(e => e.String1 != e.String1 || e.String1.Length == e.String1.Length);
```

Comment:
I get that this technically shows what we want to show, but comparing a string to itself like this is something nobody will ever write and may be confusing to readers... Maybe consider a query that could make more sense in the real world (i.e. with two different properties).

Reply:
I remember starting with different strings but wasn't happy with the sql it generated. IIRC the optimization wasn't very clear, so I decided to go with this one. Will use different strings and we can judge which one feels better.